File: LinkedList/SingleLinkedList.js
class Node{
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.length = 0;
    }

    append(data, position) {
        const node = new Node(data);

        if (position === 'front') {
            node.next = this.head;
            this.head = node;
            this.length++;
            return;
        }

        if (!this.head) {
            this.head = node;
            this.length++;
            return;
        }

        let temp = this.head;
        while(temp.next) {
            temp = temp.next;
        }
        temp.next = node;
        this.length++;
        return;
    }

    reverse() {
        let new_node = null;
        let temp = this.head;
        while (temp) {
            let next_node = temp.next;
            temp.next = new_node;
            new_node = temp;
            temp = next_node;
        }
        return new_node;
    }

    print() {
        let node = this.head;
        while(node){
            console.log(node.data);
            node = node.next;
        }
    }

    printBackwards() {
        let node = this.reverse()
        while(node){
            console.log(node.data);
            node = node.next;
        }
    }

    size() {
        return this.length;
    }

}

const main = () => {
    const ll = new LinkedList();
    const lld = new LinkedList();
    const lla = new LinkedList();
    const llf = new LinkedList();
    const arr = [6,4,3,1,10,8,9,12,0];

    for (let i = 0; i <= 5; i++) {
        ll.append(i);
        lld.append(i, position='front');
    }

    arr.forEach(i => {
        lla.append(i);
        llf.append(i, position='front')
    });

    ll.print();
    console.log()
    ll.printBackwards();
    console.log()
    console.log(ll.length);
    console.log()
    console.log(ll.size());

    console.log()

    lld.print();
    console.log()
    lld.printBackwards();
    console.log()
    console.log(lld.length);
    console.log()
    console.log(lld.size());

    console.log()

    lla.print();
    console.log()
    lla.printBackwards();
    console.log()
    console.log(lla.length);
    console.log()
    console.log(lla.size());

    console.log()

    llf.print();
    console.log()
    llf.printBackwards();
    console.log()
    console.log(llf.length);
    console.log()
    console.log(llf.size());

}

main();

File: LinkedList/SingleLinkedList.go
package main

import "fmt"

type Node struct {
	data interface{}
	next *Node
}

type LinkedList struct {
	head   *Node
	length int
}

type appendOptions struct {
	position string
}

func (list *LinkedList) append(data interface{}, options *appendOptions) {
	var node *Node = &Node{
		data: data,
		next: nil,
	}

	if options.position == "front" {
		node.next = list.head
		list.head = node
		list.length++
		return
	}

	if list.head == nil {
		list.head = node
		list.length++
		return
	}

	var temp *Node = list.head
	for temp.next != nil {
		temp = temp.next
	}
	temp.next = node
	list.length++
	return
}

func (list *LinkedList) reverse() (newNode *Node) {
	// use bubble sort for reversing
	newNode = nil
	var temp *Node = list.head
	for temp != nil {
		var next_node *Node = temp.next
		temp.next = newNode
		newNode = temp
		temp = next_node
	}
	return
}

func (list *LinkedList) print() {
	var node *Node = list.head
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
}

func (list *LinkedList) printBackwards() {
	var node *Node = list.reverse()
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
}

func (list *LinkedList) size() int {
	return list.length
}

func main() {
	var ll *LinkedList = &LinkedList{}
	var lld *LinkedList = &LinkedList{}

	for i := 0; i <= 5; i++ {
		ll.append(i, &appendOptions{})
	}

	ll.print()
	fmt.Println()
	ll.printBackwards()
	fmt.Println()
	fmt.Println(ll.length)
	fmt.Println()
	fmt.Println(ll.size())

	fmt.Println()

	for i := 0; i <= 5; i++ {
		lld.append(i, &appendOptions{position: "front"})
	}

	lld.print()
	fmt.Println()
	lld.printBackwards()
	fmt.Println()
	fmt.Println(lld.length)
	fmt.Println()
	fmt.Println(lld.size())

}
